bluetooth: fast_pair: Add crypto functions

alstrzebonski opened the pull request:
Adds crypto functions that will be used by Personalized Name Fast Pair extension. See Google Fast Pair Specification (https://developers.google.com/nearby/fast-pair/specifications/introduction) for more details. Also adds test vectors to test those functions.

Jira: NCSDK-14056

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-8128
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

NordicBuilder commented:
### Integration test specification

| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| desktop52_verification | X | | |
| test-fw-nrfconnect-ble | X | | |
| test-fw-nrfconnect-ble_samples | X | | |
| test-fw-nrfconnect-zigbee | X | | |
| test-sdk-find-my | X | | |
| test-sdk-homekit | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sub/job/test-fw-nrfconnect-nrf/job/master/24887/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

alstrzebonski commented:
I will rebase this PR on: https://github.com/nrfconnect/sdk-nrf/pull/8131 when it will be merged.

KAGA164 commented on subsys/bluetooth/services/fast_pair/include/fp_crypto.h:
```diff
@@ -103,6 +105,87 @@ size_t fp_crypto_account_key_filter_size(size_t n);
+ *
+ * @return 0 If the operation was successful. Otherwise, a (negative) error code is returned.
+ */
+int fp_crypto_aes_ctr_encrypt(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *k,
```
Maybe k->key, the parameter name will be still short and its name told us what it is exactly. Consider this also in other API
```suggestion
int fp_crypto_aes_ctr_encrypt(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *key,
```

KAGA164 replied:
You can consider also to make other parameters name more descriptive

KAGA164 commented on subsys/bluetooth/services/fast_pair/include/fp_crypto.h:
```diff
@@ -103,6 +105,87 @@ size_t fp_crypto_account_key_filter_size(size_t n);
+ *
+ * @param[out] out Buffer to receive encrypted data.
+ * @param[in] in Plaintext data.
+ * @param[in] data_len Length of data (in bytes).
```
Is it a length of `in`, `out` or both ?

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+	uint8_t aes_block[FP_CRYPTO_AES128_BLOCK_LEN];
+	int err;
+
+	if (data_len % FP_CRYPTO_AES128_BLOCK_LEN == 0) {
```
```suggestion
	if ((data_len % FP_CRYPTO_AES128_BLOCK_LEN) == 0) {
```

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+	memset(&aes_input[AES_INPUT_PAD_START], 0, AES_INPUT_PAD_LEN);
+	memcpy(&aes_input[AES_INPUT_NONCE_START], nonce, FP_CRYPTO_NONCE_LEN);
+
+	memcpy(out, in, data_len);
```
Blank line after?

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+	memcpy(&aes_input[AES_INPUT_NONCE_START], nonce, FP_CRYPTO_NONCE_LEN);
+
+	memcpy(out, in, data_len);
+	for (size_t i = 0; i < block_cnt; i++) {
```
What is a maximum value of `block_cnt`? Is it safe to cast it to `uint8_t` ?

alstrzebonski replied:
It's not specified. Fast Pair specification (see image below) casts it to `uint8_t`. I believe it shouldn't be a problem even if `block_cnt` exceeded `uint8_t` range which btw. is very unlikely. In current specification we use that function only to encrypt personalized name of the device so the name would have to exceed 256*16 characters.
![image](https://user-images.githubusercontent.com/86837740/177812052-7ddc15ee-32f8-4d9c-9bce-48bba979300f.png)

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+
+	memcpy(out, in, data_len);
+	for (size_t i = 0; i < block_cnt; i++) {
+		aes_input[0] = (uint8_t)i;
```
Blank line after?

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+
+	for (size_t i = 0; i < FP_CRYPTO_AES128_BLOCK_LEN; i++) {
+		inner_sha_input[i] = k[i] ^ HMAC_SHA_IPAD_VAL;
+	}
```
Blank line after?

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+	}
+	memset(&inner_sha_input[FP_CRYPTO_AES128_BLOCK_LEN], HMAC_SHA_IPAD_VAL,
+	       HMAC_SHA_PAD_LEN - FP_CRYPTO_AES128_BLOCK_LEN);
+	memcpy(&inner_sha_input[INNER_SHA_DATA_START], in, data_len);
```
Blank line after?

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+
+	for (size_t i = 0; i < FP_CRYPTO_AES128_BLOCK_LEN; i++) {
+		sha_input[i] = k[i] ^ HMAC_SHA_OPAD_VAL;
+	}
```
Blank line after?

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+		sha_input[i] = k[i] ^ HMAC_SHA_OPAD_VAL;
+	}
+	memset(&sha_input[FP_CRYPTO_AES128_BLOCK_LEN], HMAC_SHA_OPAD_VAL,
+	       HMAC_SHA_PAD_LEN - FP_CRYPTO_AES128_BLOCK_LEN);
```
Blank line after?

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+	int err;
+	uint8_t hmac_sha256_buf[FP_CRYPTO_SHA256_HASH_LEN];
+
+	memcpy(&out[ADDITIONAL_DATA_NONCE_START], nonce, FP_CRYPTO_NONCE_LEN);
```
Blank line after?

KAGA164 commented on subsys/bluetooth/services/fast_pair/crypto/fp_crypto_common.c:
```diff
@@ -60,3 +71,126 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
+
+size_t fp_crypto_additional_data_decoded_size(size_t packet_len)
+{
+	__ASSERT_NO_MSG(packet_len >= ADDITIONAL_DATA_ENC_DATA_START);
```
It seems to be API function, should it fails with assertion or return an error, for example 0 when packet_len is to small. In case where assertion will be disabled this function return unexpected value if packet_len will be to small. Consider adding documentation about minimum length of the packet and you can return 0 in case when packet_len is to small as a protection in case when assertion are disabled

alstrzebonski commented:
I addressed the comments and, following discussion with @MarekPieta, rebased the PR on this PR: https://github.com/nrfconnect/sdk-nrf/pull/8131 - now the first two commits are not mine, but copied from this PR. We've decided to create separate files for Fast Pair crypto for Personalized Name extension. That's where the new API has been placed. Please rereview.

shantha-14 commented:
@ns-tolu  and @guwa  has to approve before these changes are merged.

guwa reviewed (approved):
NCS BLE nightly CI passed, approve from ble verification side
https://jenkins-ncs.nordicsemi.no/blue/organizations/jenkins/latest%2Fnight%2Fsub%2Ftest-fw-nrfconnect-ble/detail/master/425/pipeline

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_crypto/include/fp_crypto_pn.h:
```diff
@@ -0,0 +1,116 @@
+ *
+ * @param[out] out 256-bit (32-byte) buffer to receive hashed result.
+ * @param[in] in Input data.
+ * @param[in] data_len Length of input data.
```
Maybe we could somehow emphasize that input data is the actual data + nonce to prevent confusion?

alstrzebonski replied:
This is just a general function to calculate HMAC-SHA256. It calculates hash from some bytes. Fast Pair personalized name extension just uses this function in some specific case. User can use `fp_crypto_additional_data_encode` that will calculate hash for him. This HMAC-SHA256 function may be useful in future. Also I want to keep it to be able to easily test it.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_pn.c:
```diff
@@ -0,0 +1,163 @@
+	int err;
+
+	if ((data_len % FP_CRYPTO_AES128_BLOCK_LEN) == 0) {
+		block_cnt = data_len / FP_CRYPTO_AES128_BLOCK_LEN;
```
I think it's unnecessary here. Why should we add brackets when it comes to "/" sign and not add them when it comes to "+" sign?

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_pn.c:
```diff
@@ -0,0 +1,163 @@
+	memcpy(out, in, data_len);
+
+	for (size_t i = 0; i < block_cnt; i++) {
+		aes_input[0] = (uint8_t)i;
```
I believe we can always cast here - even if it means cutting off the oldest bytes. See https://developers.google.com/nearby/fast-pair/specifications/characteristics#AdditionalData

alstrzebonski replied:
According to spec it would be valid and we should do the cast. It's just some index - it doesn't carry any actual data.

alstrzebonski replied:
Here is the code frome spec:
`encryptedBlock[i] = clearBlock[i] ^ AES(key, concat((uint8) i, 0x00000000000000, nonce))`

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_pn.c:
```diff
@@ -0,0 +1,163 @@
+	}
+
+	memset(&aes_input[aes_input_pad_start], 0, aes_input_pad_len);
+	memcpy(&aes_input[aes_input_nonce_start], nonce, FP_CRYPTO_PN_NONCE_LEN);
```
I don't think we need this. We have tests and also the code is straight forward. It also seems kind of problematic. The initialization ends inside for loop.

alstrzebonski commented on tests/subsys/bluetooth/fast_pair/crypto/src/main.c:
```diff
@@ -169,14 +170,118 @@ static void test_bloom_filter(void)
+		   "Error during value encryption.");
+	zassert_mem_equal(result_buf, ciphertext, sizeof(ciphertext), "Invalid encryption result.");
+
+	zassert_ok(fp_crypto_aes_ctr_decrypt(result_buf, ciphertext, sizeof(ciphertext), key,
```
It would be inconsistent with naming in other places. Nothing wrong with splitting the line for me.

MarekPieta replied:
Ok, let's leave it

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_pn.c:
```diff
@@ -0,0 +1,163 @@
+#define ADDITIONAL_DATA_NONCE_START	ADDITIONAL_DATA_SHA_LEN
+#define ADDITIONAL_DATA_ENC_DATA_START	(ADDITIONAL_DATA_NONCE_START + FP_CRYPTO_PN_NONCE_LEN)
+
+int fp_crypto_aes_ctr_encrypt(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *key,
```
Where? I couldn't find it. If data length will be set to zero the function will effectively do nothing (which is expected - if you encrypt nothing you get nothing in return). There is also no problem in encrypting data that is less than 16 bytes - actually we need this when the Personalized Name is shorter than 16 bytes.

MarekPieta replied:
Discussed offline. We can leave it as it is.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_pn.c:
```diff
@@ -0,0 +1,151 @@
+	uint8_t sha_input[hmac_sha_pad_len + FP_CRYPTO_SHA256_HASH_LEN];
+	int err;
+
+	for (size_t i = 0; i < FP_CRYPTO_AES128_BLOCK_LEN; i++) {
```
I don't think that we should multiply the existence of the same values. It's the definition of AES cypher that key is of the same length as message.

alstrzebonski replied:
Sure, I can add assertion.

alstrzebonski replied:
I will rename the variable name to `aes_key`.

alstrzebonski replied:
I added additional define as discussed in separate commit.

alstrzebonski commented on tests/subsys/bluetooth/fast_pair/crypto/src/main.c:
```diff
@@ -169,14 +172,135 @@ static void test_bloom_filter(void)
+			  "Invalid decoding result.");
+
+	/* Test data integrity check. */
+	additional_data_packet[2] += 1;
```
Ok, but honestly, what difference does it make?

MarekPieta replied:
It clearly indicates which data is test vector and reduces number of bytes allocated on stac.

MarekPieta commented:
Approved, please add the last requested changes. I start vacation from today, I would prefer not to block it for long period of time.

MarekPieta commented:
Hi @ns-tolu, @guwa
The test scope was expanded. Please rerun the nightly before merging this one to ensure we are safe.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_crypto/include/fp_crypto_pn.h:
```diff
@@ -0,0 +1,100 @@
+/*
```
Maybe we should call it fp_crypto_ext.h and make it a file that will contain all extension's functions?

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_crypto/include/fp_crypto_pn.h:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @return 0 If the operation was successful. Otherwise, a (negative) error code is returned.
+ */
+int fp_crypto_hmac_sha256(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *aes_key);
```
I wouldn't implement it in backends as soon we will have to move to PSA API anyway (https://projecttools.nordicsemi.no/jira/browse/NCSDK-15251). Actually I'm planning to do it in next few weeks.

kapi-no replied:
IMO, PSA API should be implemented as another crypto backend because it does not have the necessary crypto operations for all Fast Pair extension combinations. If PSA API does not support HMAC SHA256, you can move the implementation from the fp_crypto_pn.c. By having a custom implementation using SHA256, you limit the performance for other backends that support this crypto operation.

MarekPieta replied:
Discussed offline. Will be considered as a separate follow-up task (if we introduce a backend with native support of this crypto operations we could e.g. make these functions weak)

alstrzebonski replied:
https://projecttools.nordicsemi.no/jira/browse/NCSDK-16280 here is a task for it.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_crypto/include/fp_crypto_pn.h:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @return 0 If the operation was successful. Otherwise, a (negative) error code is returned.
+ */
+int fp_crypto_aes_ctr_encrypt(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *key,
```
And those functions could end up in `fp_crypto_ext_priv.h` so that we can test it but don't expose it.

kapi-no replied:
Hmm, couldn't you include the source file (include "fp_crypto_pn.c:") in the test source file to avoid unnecessary private headers?

alstrzebonski replied:
Currently we are building crypto and storage as a separate library to avoid duplicating KConfig options. Then the libraries are linked. The `_priv` approach has been already used in Fast Pair storage. Thanks to this we can explicitle state what API is needed for tests and what API is is actually used by Fast Pair service.

KAGA164 commented on subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto:
```diff
@@ -37,6 +37,11 @@ config BT_FAST_PAIR_CRYPTO_MBEDTLS
 endchoice
 
+config BT_FAST_PAIR_CRYPTO_EXT_PN
+	bool "Fast Pair crypto for Personalized Name extension"
```
I would say that it is nice to keep prompt as short as possible. You can remove Fast Pair suffix for all prompt in this file. For example:

```suggestion
	bool Crypto for Personalized Name extension"
```

KAGA164 reviewed (commented):
Looks good. Just added one comment

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_crypto/include/fp_crypto.h:
```diff
@@ -103,6 +109,72 @@ size_t fp_crypto_account_key_filter_size(size_t n);
 int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *account_key_list,
 				 size_t n, uint8_t salt);
 
+/** Encrypt data using AES-CTR.
+ *
+ * @param[out] out Buffer to receive encrypted data.
+ * @param[in] in Plaintext data.
+ * @param[in] data_len Length of input and output data (in bytes).
+ * @param[in] key 128-bit (16-byte) AES key.
+ * @param[in] nonce 64-bit (8-byte) nonce.
+ *
+ * @return 0 If the operation was successful. Otherwise, a (negative) error code is returned.
+ */
+int fp_crypto_aes_ctr_encrypt(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *key,
+			      const uint8_t *nonce);
+
+/** Decrypt data using AES-CTR.
+ *
+ * @param[out] out Buffer to receive plaintext data.
+ * @param[in] in Ciphertext data.
+ * @param[in] data_len Length of input and output data (in bytes).
+ * @param[in] key 128-bit (16-byte) AES key.
+ * @param[in] nonce 64-bit (8-byte) nonce used to encrypt data.
+ *
+ * @return 0 If the operation was successful. Otherwise, a (negative) error code is returned.
+ */
+int fp_crypto_aes_ctr_decrypt(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *key,
+			      const uint8_t *nonce);
```
Please move this part next to the fp_crypto_aes128_encrypt / decrypto functions.

Please indicated the operation size as for previous AES functions (e.g. fp_crypto_aes128_ctr_decrypt)

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_crypto/include/fp_crypto.h:
```diff
@@ -103,6 +109,72 @@ size_t fp_crypto_account_key_filter_size(size_t n);
+ *
+ * @return 0 If the operation was successful. Otherwise, a (negative) error code is returned.
+ */
+int fp_crypto_hmac_sha256(uint8_t *out, const uint8_t *in, size_t data_len, const uint8_t *aes_key);
```
Please move this function below fp_crypto_sha256() as they are in the same category (hashing functions)

kapi-no commented on tests/subsys/bluetooth/fast_pair/crypto/Kconfig:
```diff
@@ -11,3 +11,8 @@ endmenu
 source "Kconfig.zephyr"
 endmenu
+
+config BT_FAST_PAIR_TEST_CRYPTO_EXT
+	bool "Test crypto for Fast Pair extensions"
+	help
+	  Test crypto functions for Fast Pair extensions.
```
I think this is no longer necessary as crypto implementations for extensions are resolved at the link time (If the test project in the main.c file references a given crypto API it should not be optimized by the linker)

kapi-no commented on tests/subsys/bluetooth/fast_pair/crypto/src/main.c:
```diff
@@ -169,14 +170,136 @@ static void test_bloom_filter(void)
 
+static void test_aes_ctr(void)
+{
+	if (!IS_ENABLED(CONFIG_BT_FAST_PAIR_TEST_CRYPTO_EXT)) {
+		ztest_test_skip();
+		return;
+	}
```
This is no longer necessary. Let's test everything each time for each crypto backend (no need for another job in the testcase.yaml file). Same for other functions

alstrzebonski replied:
Ok, we can simplify it that way. @MarekPieta do you agree?

kapi-no commented on tests/subsys/bluetooth/fast_pair/crypto/src/main.c:
```diff
@@ -169,14 +170,136 @@ static void test_bloom_filter(void)
 			  "Invalid resulting filter.");
 }
 
+static void test_aes_ctr(void)
```
Please arrange the testing order according to the fmn_crypto.h

1. sha256
2. hmac_sha256
3. aes128
4. aes128_ctr
(...)

kapi-no commented on tests/subsys/bluetooth/fast_pair/crypto/testcase.yaml:
```diff
@@ -18,3 +18,16 @@ tests:
       - nrf9160dk_nrf9160_ns
       - qemu_cortex_m3
     extra_args: CONF_FILE=prj_tinycrypt.conf
+  fast_pair.crypto.mbedtls.core:
+    platform_allow:
+      nrf52840dk_nrf52840
+    integration_platforms:
+      - nrf52840dk_nrf52840
+    extra_args: CONFIG_BT_FAST_PAIR_TEST_CRYPTO_EXT=n
+  fast_pair.crypto.tinycrypt.core:
+    platform_allow:
+      nrf52840dk_nrf52840 qemu_cortex_m3
+    integration_platforms:
+      - nrf52840dk_nrf52840
+      - qemu_cortex_m3
+    extra_args: CONF_FILE=prj_tinycrypt.conf CONFIG_BT_FAST_PAIR_TEST_CRYPTO_EXT=n
```
No need for this test definition if you agree with my previous review comments

MarekPieta replied:
Discussed with @alstrzebonski. We can simplify the configuration of both fp_crypto library and the unit test and rely on optimizing out unused functions (I noticed that memory consumption is not increased, so we should be fine).

kapi-no reviewed (approved):
Looking good now. IMO, it would be great to arrange the test functions according to the fp_crypto.h (both their implementation order and order inside the ```ztest_test_suite``` function). See my previous comment:

https://github.com/nrfconnect/sdk-nrf/pull/8128#discussion_r935590692

Please consider doing the same for fp_crypto_common.c (hmac_sha256 should be the first and then according to the order in the header)

alstrzebonski commented:
Rebased on main.

alstrzebonski commented:
Applied last review comments.

alstrzebonski commented:
Rebased on main again.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_common.c:
```diff
@@ -63,3 +144,68 @@ int fp_crypto_account_key_filter(uint8_t *out, const struct fp_account_key *acco
 	return 0;
 }
+
+int fp_crypto_aes128_ctr_decrypt(uint8_t *out, const uint8_t *in, size_t data_len,
```
Move the decrypt right below encrypt, please

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_common.c:
```diff
@@ -15,6 +16,86 @@ LOG_MODULE_REGISTER(fp_crypto, CONFIG_FP_CRYPTO_LOG_LEVEL);
+#define ADDITIONAL_DATA_SHA_LEN		8
+#define ADDITIONAL_DATA_NONCE_POS	(ADDITIONAL_DATA_SHA_POS + ADDITIONAL_DATA_SHA_LEN)
+#define ADDITIONAL_DATA_ENC_DATA_POS	(ADDITIONAL_DATA_NONCE_POS + FP_CRYPTO_PN_NONCE_LEN)
+
```
Could you please use 2 empty lines between data and first function implementation?

alstrzebonski commented:
Hi @ns-tolu, @guwa
Please rerun the nightly and we are ready to merge.

KAGA164 reviewed (approved):
Looks good
